feat: new SAP SQL Anywhere database driver

bhoupert opened the pull request:
Hello,

This is my first pull request.
I have developped a new CI4 database driver for SAP SQL Anywhere (merged from SQL Server CI4 driver).

I have changed some things in these files :
- `app/Config/Database.php` : added two new connection parameters (`'engine'` and `'conAuth'`) as SQL Anywhere uses both database and engine (an engine can run several databases), and can be used in "authentication mode" (as we need a specific connection string)
- `system/Database/BaseConnection.php` and `system/Database/BaseConnection.php` : to retrieve the two new vars

The new directory `system/Database/SASQL` embeds all the driver's code

I have tested each function, based on CI4 documentation (https://codeigniter.com/user_guide/database/index.html) and everything is working fine.

**Checklist:**
- [ ] Securely signed commits
- [ ] Component(s) with PHPDoc blocks, only if necessary or adds value
- [ ] Unit testing, with >80% coverage
- [ ] User guide updated
- [ ] Conforms to style guide

kenjis commented:
Thank you for great contributing!

You must [GPG-sign](https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/signing.md) your work, certifying that you either wrote the work or otherwise have the right to pass it on to an open-source project. See [Developer's Certificate of Origin](https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/DCO.md).
See https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/pull_request.md#signing

There are some GitHub Actions errors. Please fix all of them.
Read:
- https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/pull_request.md#php-style
- https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/pull_request.md#static-analysis-on-php-code
- https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/pull_request.md#unit-testing

kenjis commented:
To test this PR, we need SAP SQL Anywhere serivce in GitHub Action workflow.
Does anyone know about it?

kenjis commented:
This is enhancement. So it should be sent to the next minor version branch `4.4`.
I changed the base branch.
See https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/pull_request.md#branching

bhoupert commented:
Hi, 


Thank you for replying. 
You just need no know that SAP SQL Anywhere is under commercial licence, so maybe it will we be difficult to integrate it. Besides, we need to embed a specific DLL in php.ini to access specific functions. 
I'm gonna try to pass the tests, but it's gonna be difficult. In example, phpstan retrieve errors because it doesn't know about many functions and constants provided by the SQL Anywhere DLL : 





-- 

Bien cordialement, 

Baptiste HOUPERT 
__________________________ 

***@***.*** 
Tél. : 06.86.97.65.32 
----- Mail original -----

De: "kenjis" ***@***.***> 
À: "codeigniter4/CodeIgniter4" ***@***.***> 
Cc: "bhoupert" ***@***.***>, "Author" ***@***.***> 
Envoyé: Samedi 14 Janvier 2023 00:35:41 
Objet: Re: [codeigniter4/CodeIgniter4] Contribution : new SAP SQL Anywhere database driver for CodeIgniter 4 (PR #7114) 



To test this PR, we need SAP SQL Anywhere serivce in GitHub Action workflow. 
Does anyone know about it? 
— 
Reply to this email directly, view it on GitHub , or unsubscribe . 
You are receiving this because you authored the thread. Message ID: <codeigniter4/CodeIgniter4/pull/7114/c1382585090 @ github . com>

kenjis commented:
> You just need no know that SAP SQL Anywhere is under commercial licence, so maybe it will we be difficult to integrate it.

Oh, it is a really bad news. If we cannot build CI environment, it is difficult to merge into CodeIgniter4.
We don't have SAP SQL Anywhere license, so we cannot test at all.
It might be better that you release SAP SQL Anywhere Driver package for CodeIgniter4.

> I'm gonna try to pass the tests, but it's gonna be difficult. In example, phpstan retrieve errors because it doesn't know about many functions and constants provided by the SQL Anywhere DLL :

You can run it in your environment with the DLL.

bhoupert commented:
I guess it was the same for SQL Server ?
I can release it in a single package but I even need to make some changes in the few core files : I will make a new pull request for that.

kenjis commented on system/Database/BaseConnection.php:
```diff
@@ -74,6 +76,20 @@ abstract class BaseConnection implements ConnectionInterface
+     *
+     * @var string
+     */
+    protected $engine;
```
This property is only for SASQL. BaseConnection should not have it.

kenjis commented on system/Database/BaseConnection.php:
```diff
@@ -74,6 +76,20 @@ abstract class BaseConnection implements ConnectionInterface
+     *
+     * @var string
+     */
+    protected $conAuth;
```
This property is only for SASQL. BaseConnection should not have it.

kenjis commented on system/Database/Database.php:
```diff
@@ -99,6 +99,8 @@ protected function parseDSN(array $params): array
             'DBDriver' => $dsn['scheme'],
             'hostname' => isset($dsn['host']) ? rawurldecode($dsn['host']) : '',
+            'engine'   => isset($dsn['engine']) ? rawurldecode($dsn['engine']) : '',
+            'conAuth'  => isset($dsn['conAuth']) ? rawurldecode($dsn['conAuth']) : '',
```
These items are only for SASQL. Database.php should not have it.

kenjis commented on app/Config/Database.php:
```diff
@@ -27,6 +27,8 @@ class Database extends Config
         'DSN'      => '',
         'hostname' => 'localhost',
+        'engine'   => '',
+        'conAuth'  => '',
```
The default config is for MySQL. These items are not needed.

kenjis commented on app/Config/Database.php:
```diff
@@ -51,6 +53,8 @@ class Database extends Config
         'DSN'         => '',
         'hostname'    => '127.0.0.1',
+        'engine'      => '',
+        'conAuth'     => '',
```
The test config is sample for SQLite. These items are not needed.

kenjis commented:
@bhoupert Very nice!

You are using our vendor name `CodeIgniter` in the namespaces of your code, 
but you can use your own namespace for the driver.
See https://forum.codeigniter.com/thread-78294-post-383813.html

So if you use your vendor name, you could publish it as a Composer package,
and users can install it outside of `system` folder, which will be replaced by upgrading.

bhoupert commented:
Thank you @kenjis .
I have :
- moved files to /app/directory
- renamed the driver folder name
- changed namespaces
- update README.md

It works fine 👍
